core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java

```diff
@@ -0,0 +1,199 @@
+ */
+public class GlobalCheckpointService extends AbstractIndexShardComponent {
+
+    /**
+     * This map holds the last known local checkpoint for every shard copy that's active.
+     * All shard copies in this map participate in determining the global checkpoint
+     * keyed by allocation ids
+     */
+    final ObjectLongMap<String> activeLocalCheckpoints;
+
+    /**
+     * This map holds the last known local checkpoint for every initializing shard copy that's has been brought up
+     * to speed through recovery. These shards are treated as valid copies and participate in determining the global checkpoint.
+     * Keyed by allocation ids
+     */
+    final ObjectLongMap<String> inSyncLocalCheckpoints; // keyed by allocation ids
+
+
+    /**
+     * This map holds the last known local checkpoint for every initializing shard copy that is still undergoing recovery.
+     * These shards <strong>do not</strong> participate in determining the global checkpoint. These map is needed to make sure that when
+     * shards are promoted to {@link #inSyncLocalCheckpoints} we use the highest known checkpoint, even if we index concurrently
+     * while recovering the shard.
+     * Keyed by allocation ids
+     */
+    final ObjectLongMap<String> trackingLocalCheckpoint;  // keyed by allocation ids
+
+    long globalCheckpoint = SequenceNumbersService.UNASSIGNED_SEQ_NO;
+
+    public GlobalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        activeLocalCheckpoints = new ObjectLongHashMap<>(1 + indexSettings.getNumberOfReplicas());
+        inSyncLocalCheckpoints = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
+        trackingLocalCheckpoint = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
+    }
+
+
+    /**
+     * notifies the service of a local checkpoint. if the checkpoint is lower than the currently known one,
+     * this is a noop. Last, if the allocation id is not yet known, it is added to {@link #trackingLocalCheckpoint}
+     */
+    synchronized public void updateLocalCheckpoint(String allocationId, long localCheckpoint) {
+        if (updateLocalCheckPointInMap(allocationId, localCheckpoint, activeLocalCheckpoints, "active")) {
+            return;
+        }
+        if (updateLocalCheckPointInMap(allocationId, localCheckpoint, inSyncLocalCheckpoints, "inSync")) {
+            return;
+        }
+        if (updateLocalCheckPointInMap(allocationId, localCheckpoint, trackingLocalCheckpoint, "tracking")) {
+            return;
+        }
+        trackingLocalCheckpoint.put(allocationId, localCheckpoint);
+        logger.trace("added local checkpoint of [{}] ([{}]) to the tracking map", allocationId, localCheckpoint);
+    }
+
+    private boolean updateLocalCheckPointInMap(String allocationId, long localCheckpoint, ObjectLongMap<String> checkpointsMap, String name) {
+        int indexOfKey = checkpointsMap.indexOf(allocationId);
```

Comment:
Assert that the current thread holds a lock on `this`? The results from `ObjectLongMap#indexOf` remain valid only if no one else is mutating.